File: dp/c/compensating_transaction.c
// Compensating Transaction Pattern in C

#include <stdio.h>
#include <stdbool.h>

typedef bool (*Action)(void);
typedef void (*Compensation)(void);

typedef struct { const char* name; Action action; Compensation compensate; } Step;

#define MAX_STEPS 16

bool execute_transaction(Step* steps, int n) {
    int completed = 0;
    for (int i = 0; i < n; i++) {
        printf("[TX] Executing: %s\n", steps[i].name);
        if (steps[i].action()) {
            completed++;
        } else {
            printf("[TX] FAILED at '%s'\n", steps[i].name);
            for (int j = completed - 1; j >= 0; j--) {
                printf("[TX] Compensating: %s\n", steps[j].name);
                steps[j].compensate();
            }
            return false;
        }
    }
    printf("[TX] All steps completed.\n");
    return true;
}

bool debit(void) { printf("  Debited $100\n"); return true; }
void credit_back(void) { printf("  Credited $100 back\n"); }
bool reserve(void) { printf("  Reserved 2 items\n"); return true; }
void release(void) { printf("  Released items\n"); }
bool ship(void) { printf("  Shipped\n"); return true; }
void cancel_ship(void) { printf("  Cancelled shipment\n"); }
bool fail_pay(void) { return false; }
void refund(void) { printf("  Refunded\n"); }

int main(void) {
    printf("=== Success ===\n");
    Step s1[] = {
        {"Debit", debit, credit_back},
        {"Reserve", reserve, release},
        {"Ship", ship, cancel_ship},
    };
    execute_transaction(s1, 3);

    printf("\n=== Failure ===\n");
    Step s2[] = {
        {"Debit", debit, credit_back},
        {"Payment", fail_pay, refund},
        {"Ship", ship, cancel_ship},
    };
    execute_transaction(s2, 3);
    return 0;
}

File: dp/rust/src/compensating_transaction.rs
// Compensating Transaction Pattern in Rust
// Undo completed steps when a multi-step operation fails

use std::fmt;

struct Transaction {
    steps: Vec<Step>,
}

struct Step {
    name: String,
    action: Box<dyn Fn() -> Result<(), String>>,
    compensation: Box<dyn Fn()>,
}

impl Transaction {
    fn new() -> Self {
        Transaction { steps: Vec::new() }
    }

    fn step<A, C>(mut self, name: &str, action: A, compensation: C) -> Self
    where
        A: Fn() -> Result<(), String> + 'static,
        C: Fn() + 'static,
    {
        self.steps.push(Step {
            name: name.to_string(),
            action: Box::new(action),
            compensation: Box::new(compensation),
        });
        self
    }

    fn execute(&self) -> Result<(), TransactionError> {
        let mut completed = Vec::new();

        for step in &self.steps {
            println!("[TX] Executing: {}", step.name);
            match (step.action)() {
                Ok(()) => {
                    completed.push(step);
                }
                Err(e) => {
                    println!("[TX] FAILED at '{}': {}", step.name, e);
                    println!("[TX] Compensating {} completed steps...", completed.len());

                    for step in completed.iter().rev() {
                        println!("[TX] Compensating: {}", step.name);
                        (step.compensation)();
                    }

                    return Err(TransactionError {
                        step: step.name.clone(),
                        cause: e,
                        compensated: completed.iter().map(|s| s.name.clone()).collect(),
                    });
                }
            }
        }

        println!("[TX] All steps completed successfully.");
        Ok(())
    }
}

#[derive(Debug)]
struct TransactionError {
    step: String,
    cause: String,
    compensated: Vec<String>,
}

impl fmt::Display for TransactionError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Transaction failed at '{}': {}. Compensated: {:?}",
            self.step, self.cause, self.compensated
        )
    }
}

fn main() {
    // Successful transaction
    println!("=== Success Case ===");
    let tx = Transaction::new()
        .step("Debit account", || { println!("  Debited $100"); Ok(()) }, || println!("  Credited $100 back"))
        .step("Reserve stock", || { println!("  Reserved 2 items"); Ok(()) }, || println!("  Released 2 items"))
        .step("Send email", || { println!("  Email sent"); Ok(()) }, || println!("  (email can't be unsent)"));
    let _ = tx.execute();

    // Failing transaction
    println!("\n=== Failure Case ===");
    let tx = Transaction::new()
        .step("Debit account", || { println!("  Debited $50"); Ok(()) }, || println!("  Credited $50 back"))
        .step("Reserve stock", || Err("Out of stock".into()), || println!("  Released stock"))
        .step("Send email", || { println!("  Email sent"); Ok(()) }, || println!("  (not reached)"));

    if let Err(e) = tx.execute() {
        println!("\n{}", e);
    }
}
